### ISSUE TITLE
Problem in creating export through dbus

### ISSUE THREAD

**Issue by r-deshmukh** [nfs-ganesha/nfs-ganesha#818]
It is observed that when we create export using dbus with export id 1 then we are not able to add another export. The command which tried to add export returning with following error :
 "check_export_duplicate :EXPORT :DEBUG :Export 1 already exists"

Looks like following code added this error message: 
-----------------------------------------------------------------------
commit 7c44cb498a5c759414e74531314b22596ce55fab
Author: Luo.muyao <luo.muyao@zte.com.cn>
Date:   Thu Jun 3 14:59:28 2021 +0800

    EXPORT: fixed issue where adding duplicate ExportID failed by DBus

    When use CEPH_FSAL as backend, if we add existing ExportID by DBus,
    it will cause the active EXPORT to disconnect from the MDS, and occur
    -108 error.

    Each EXPORT is a CephFS client and set different UUID, the same ExportID
    has the same UUID, after fail to add an existing ExportID and goes to
    the release duplicate EXPORT process, this triggers MDS to disconnect
    all client connections that have the UUID, include active export.

    Change-Id: If5a7147b5b63714b339774dac1fcbaaac806ef42
    Signed-off-by: Luo.muyao <luo.muyao@zte.com.cn>
-----------------------------------------------------------------------

When tried removing above fix, we are able to create the export OR it fixed the problem. Please review and let me know if more information is needed.

**Comment by ffilz** [comment]
Are you assigning each export a unique export_id? The code in the patch is very simple and just checks each export that is being created by dbus to assure that the export_id is not already registered. What is your original export config? What is the command you have used to add exports? Are you able to add even one export with dbus?

**Comment by r-deshmukh** [comment]
Actually this is bit strange problem. 
When I add an export with export id 1 then I am not able to add another export with say export id 2 or 3. 
But when I skip adding export with export id 1 then there is no problem for adding another export with say export id 2 or 3.

I am using dbus command to create export and using VFS FSAL

Command used is as : 
# Add export with export id 1
$> dbus-send --print-reply --system --dest=org.ganesha.nfsd /org/ganesha/nfsd/ExportMgr org.ganesha.nfsd.exportmgr.AddExport string:/tmp/nfs_export1.conf string:'EXPORT(export_id=1)'
method return time=1653043086.153179 sender=:1.46 -> destination=:1.77 serial=33 reply_serial=2
   string "1 exports added"

$>cat  /tmp/nfs_export1.conf
EXPORT
{
    Export_Id = 1;
    Path = "/mnt/lustre/mac1011/nfs_test1";
    Pseudo = "/mnt/lustre/mac1011/nfs_test1";
    Protocols = 3, 4;
    Access_Type = RW;
    Squash = No_root_squash;
    FSAL {
           Name = VFS;
    }
}
 
# Add export with export id 2
 dbus-send --print-reply --system --dest=org.ganesha.nfsd /org/ganesha/nfsd/ExportMgr org.ganesha.nfsd.exportmgr.AddExport string:/tmp/nfs_export2.conf string:'EXPORT(export_id=2)'
**_Error org.freedesktop.DBus.Error.InvalidFileContent: Selected entries in /tmp/nfs_export2.conf already active!!!_**

$>cat /tmp/nfs_export2.conf
EXPORT
{
    Export_Id = 2;
    Path = "/mnt/lustre/mac1011/nfs_test2";
    Pseudo = "/mnt/lustre/mac1011/nfs_test2";
    Protocols = 3, 4;
    Access_Type = RW;
    Squash = No_root_squash;
    FSAL {
           Name = VFS;
    }
}
I hope this is helpful. Sorry for bad format.

**Comment by ffilz** [comment]
I don't use dbus to manage exports, so I'm not sure if this is unique to dbus. I even tried adding that check to the update export that is used for regular config read and it didn't cause a problem (it would fail the update though...). Could you enable export debug:

LOG {
    COMPONENTS {
         EXPORT=FULL_DEBUG;
    }
}

And try a failing case. Maybe from that I can determine where things are going wrong.

Thanks

**Comment by martin-schwenke** [comment]
I think I've got this.   I haven't tested the fix, but it seems obvious.  The word "DEFAULTS" seems to be the give-away...  :wink: 

The above commit adds the call of `u.blk.check()` after `do_block_init()`, which seems to set the defaults.  The default value for `Export_id` is `1`, so any time this code gets called when `Export_id` `1` already exists, it will fail.

Shouldn't `u.blk.check()` be called after after `do_block_load()`, which actually parses the values from wherever (`node`?) into `param_struct`?  Like this:

  https://review.gerrithub.io/c/ffilz/nfs-ganesha/+/538588

I assume it is the failed call to `u.blk.commit()` that causes the problem in `FSAL_CEPH`?

I also feel inclined to change the sense of the result of the `.check()` function, which surely should return `true` if the check passes.

**Comment by martin-schwenke** [comment]
Tested this fix.  Seems to work.  No longer get spurious errors about duplicates.  Duplicates are still correctly detected.

**Comment by ffilz** [comment]
A fix for this is in V4.0.6 please verify and close this issue if you find it resolved.

**Comment by martin-schwenke** [comment]
On V4.0.6:
```
# ganesha.nfsd -v
NFS-Ganesha Release = V4.0.6
nfs-ganesha compiled on Jun 27 2022 at 23:10:56
Release comment = GANESHA file server is 64 bits compliant and supports NFS v3,4.0,4.1 (pNFS) and 9P
Git HEAD = 5339a10b55b0672c7029ce192914141034d89d4a
Git Describe = V4.0.6-0-g5339a10b5
# ganesha.nfsd
# dbus-send --print-reply --system --dest=org.ganesha.nfsd /org/ganesha/nfsd/ExportMgr org.ganesha.nfsd.exportmgr.AddExport string:/tmp/nfs_export1.conf string:'EXPORT(export_id=1)'
method return time=1656371599.632665 sender=:1.20 -> destination=:1.21 serial=7 reply_serial=2
   string "1 exports added"
# dbus-send --print-reply --system --dest=org.ganesha.nfsd /org/ganesha/nfsd/ExportMgr org.ganesha.nfsd.exportmgr.AddExport string:/tmp/nfs_export2.conf string:'EXPORT(export_id=2)'
method return time=1656371602.517525 sender=:1.20 -> destination=:1.22 serial=11 reply_serial=2
   string "1 exports added"
```
While on V4.0.5:
```
# ganesha.nfsd -v
NFS-Ganesha Release = V4.0.5
nfs-ganesha compiled on Jun 27 2022 at 23:07:50
Release comment = GANESHA file server is 64 bits compliant and supports NFS v3,4.0,4.1 (pNFS) and 9P
Git HEAD = c7a1fc5378c6c9a900df88592d683ad39bb9910a
Git Describe = V4.0.5-0-gc7a1fc537
# ganesha.nfsd
# dbus-send --print-reply --system --dest=org.ganesha.nfsd /org/ganesha/nfsd/ExportMgr org.ganesha.nfsd.exportmgr.AddExport string:/tmp/nfs_export1.conf string:'EXPORT(export_id=1)'
method return time=1656371518.015255 sender=:1.17 -> destination=:1.18 serial=8 reply_serial=2
   string "1 exports added"
# dbus-send --print-reply --system --dest=org.ganesha.nfsd /org/ganesha/nfsd/ExportMgr org.ganesha.nfsd.exportmgr.AddExport string:/tmp/nfs_export2.conf string:'EXPORT(export_id=2)'
Error org.freedesktop.DBus.Error.InvalidFileContent: Selected entries in /tmp/nfs_export2.conf already active!!!
```
So fixed in V4.0.6.

I'm not the issue reporter, so don't know if I should close.  @r-deshmukh, are you happy to close?  Thanks...

**Comment by r-deshmukh** [comment]
> @r-deshmukh, are you happy to close?
Ok, let me close this. Thanks!
